src/System.Private.CoreLib/shared/System/Runtime/InteropServices/GCHandle.cs

```diff
@@ -0,0 +1,189 @@
+
+        /// <summary>Creates a new GC handle for an object.</summary>
+        /// <param name="value">The object that the GC handle is created for.</param>
+        /// <returns>A new GC handle that protects the object.</returns>
+        public static GCHandle Alloc(object value) => new GCHandle(value, GCHandleType.Normal);
+
+        /// <summary>Creates a new GC handle for an object.</summary>
+        /// <param name="value">The object that the GC handle is created for.</param>
+        /// <param name="type">The type of GC handle to create.</param>
+        /// <returns>A new GC handle that protects the object.</returns>
+        public static GCHandle Alloc(object value, GCHandleType type) => new GCHandle(value, type);
+
+        /// <summary>Frees a GC handle.</summary>
+        public void Free()
+        {
+            // Free the handle if it hasn't already been freed.
+            IntPtr handle = Interlocked.Exchange(ref m_handle, IntPtr.Zero);
+            ThrowIfInvalid(handle);
+            InternalFree(GetHandleValue(handle));
+        }
+
+        // Target property - allows getting / updating of the handle's referent.
+        public object Target
+        {
+            get
+            {
+                ThrowIfInvalid();
+                return InternalGet(GetHandleValue());
+            }
+            set
+            {
+                ThrowIfInvalid();
+
+                if (IsPinned && !Marshal.IsPinnable(value))
+                {
+                    ThrowHelper.ThrowArgumentException(ExceptionResource.ArgumentException_NotIsomorphic, ExceptionArgument.value);
+                }
+
+                InternalSet(GetHandleValue(), value);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve the address of an object in a Pinned handle.  This throws
+        /// an exception if the handle is any type other than Pinned.
+        /// </summary>
+        public IntPtr AddrOfPinnedObject()
+        {
+            // Check if the handle was not a pinned handle.
+            // You can only get the address of pinned handles.
+            ThrowIfInvalid();
+            if (!IsPinned)
+            {
+                ThrowHelper.ThrowInvalidOperationException_HandleIsNotPinned();
+            }
+
+            // Get the address.
+
+            object target = InternalGet(GetHandleValue());
+            if (target is null)
+            {
+                return default;
+            }
+
+            unsafe
+            {
+                if (RuntimeHelpers.ObjectHasComponentSize(target))
+                {
+                    if (target.GetType() == typeof(string))
+                    {
+                        return (IntPtr)Unsafe.AsPointer(ref Unsafe.As<string>(target).GetRawStringData());
+                    }
+
+                    Debug.Assert(target is Array);
+                    return (IntPtr)Unsafe.AsPointer(ref Unsafe.As<Array>(target).GetRawArrayData());
```

Comment:
@jkotas, I'm still seeing around a ~15% hit here for AddrOfPinnedObject on a pinned array.  I'm not entirely sure why, as it seems this shouldn't be doing significantly more work than in the array path prior to these changes, and the string path actually improves a bit over where it is today.  I tried removing the VALIDATEOBJECT call from GetRawArrayData just to see if it would have any effect, but it didn't seem to (it appears to be mostly debug logic).  Suggestions?

Reply:
There were some instruction-level inefficiencies contributing to the regression. I have streamline the code to help the JIT with generating good code. `AddrOfPinnedObject` on array is faster than what is in master for me now.

`GetRawArrayData` is still incurs a call overhead, but I would not worry about it. I expect that we will make `GetRawArrayData` to be implemented in C# at some point (once we are ready to do this: https://github.com/dotnet/coreclr/issues/18111) and it can be fully inlined at that point.

Reply:
Nice. Thank you